src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,222 @@
+    if there is such a pair, or an arbitrary, valid page span if there is not.
+-}
+search :: SerialisedKey -> IndexOrdinary -> PageSpan
+search key (IndexOrdinary lastKeys) = PageSpan (PageNo start) (PageNo end) where
+
+    start :: Int
+    start | protoStart < pageCount = protoStart
+          | otherwise              = maybe 0 succ $
+                                     findIndexR (/= lastKeys ! pred pageCount) $
+                                     lastKeys
+
+    end :: Int
+    end | protoStart < pageCount
+            = start + length (takeWhile (== key) (drop (succ start) lastKeys))
+        | otherwise
+            = pred pageCount
+
+    protoStart :: Int
+    protoStart = searchForProtoStart 0 (length lastKeys)
+
+    searchForProtoStart :: Int -> Int -> Int
+    searchForProtoStart first count
+        | count == 0               = first
+        | key <= lastKeys ! center = searchForProtoStart first frontCount
+        | otherwise                = searchForProtoStart (succ center) rearCount
+        where
+
+        frontCount, rearCount, center :: Int
+        frontCount = count `div` 2
+        rearCount  = pred (count - frontCount)
+        center     = first + frontCount
+
+    pageCount :: Int
+    pageCount = length lastKeys
+
+-- * Deserialisation
+
+{-|
+    Reads an index along with the number of entries of the respective run from a
+    byte string.
+
+    The byte string must contain the serialised index exactly, with no leading
+    or trailing space.
+-}
+fromSBS :: ShortByteString -> Either String (NumEntries, IndexOrdinary)
+fromSBS shortByteString@(ShortByteString byteArray)
+    | fullSize < 12
+        = Left "Doesn't contain header and footer"
+    | version == byteSwap32 supportedVersion
+        = Left "Non-matching endianness"
+    | version /= supportedVersion
+        = Left "Unsupported version"
+    | otherwise
+        = (,) <$> entriesCount <*> index
+    where
+
+    fullSize :: Int
+    fullSize = ShortByteString.length shortByteString
+
+    fullBytes :: Primitive.Vector Word8
+    fullBytes = Primitive.Vector 0 fullSize byteArray
+
+    version :: Word32
+    version = indexByteArray byteArray 0
+
+    postVersionBytes :: Primitive.Vector Word8
+    postVersionBytes = Primitive.drop 4 fullBytes
+
+    lastKeysBytes, entriesCountBytes :: Primitive.Vector Word8
+    (lastKeysBytes, entriesCountBytes)
+        = Primitive.splitAt (fullSize - 12) postVersionBytes
+
+    entriesCount :: Either String NumEntries
+    entriesCount
+        | (fromIntegral asWord64 :: Integer) > fromIntegral (maxBound :: Int)
+            = Left "Number of entries not representable as Int"
+        | otherwise
+            = Right (NumEntries (fromIntegral asWord64))
+        where
+
+        asWord64 :: Word64
+        asWord64 = indexByteArray entriesCountRep 0
+
+        entriesCountRep :: ByteArray
+        Primitive.Vector _ _ entriesCountRep = Primitive.force entriesCountBytes
```

Comment:
I don't think this `force` is required

Reply:
I think it is. 😉 The thing is that the byte array underlying `entriesCountBytes` can comprise more than just the bytes that represent the number of entries. In fact, this is always the case here, because `entriesCountBytes` is obtained from `fullBytes` by slicing and thus the byte array is the one that contains the whole byte sequence that is deserialized.

Reply:
`entiresCountRep` is only used in `asWord64`, and the byte array is not kept alive after that, that's why I think it's not necessary

Reply:
It’s not about resource usage but about correctness: The array that `entriesCountRep` contains is the original array, which contains the whole serialized index, and the fact that `entriesCountRep` spans only a small part of it comes from the other two fields of the primitive vector saying so. These other two fields we lose when converting a primitive vector into an array via pattern matching, and for the result to be correct we have to prune the array before the conversion, which we do using `force`.

Reply:
I see, I was thinking about it wrong